harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -116,6 +171,127 @@ def _set_random_seeds(cls: Type["PyTorchTrialController"], seed: int) -> None:
         # torch.backends.cudnn.deterministic = True
         # torch.backends.cudnn.benchmark = False
 
+    def report_training_metrics(self, steps_completed):
+        metrics = self._prepare_metrics(self.training_metrics)
+        metrics = self.context.distributed.broadcast(metrics)
+
+        for callback in self.callbacks.values():
+            callback.on_training_workload_end(
+                avg_metrics=metrics["avg_metrics"],
+                batch_metrics=metrics["batch_metrics"],
+            )
+
+        # Only report on the chief worker
+        if self.is_chief:
+            avg_metrics = metrics.get("avg_metrics", {})
+            batch_metrics = metrics.get("batch_metrics", [])
+
+            self.metric_writer.on_train_step_end(
+                self.steps_completed,
+                avg_metrics,
+                batch_metrics,
+            )
+            self.core_context.train.report_training_metrics(
+                steps_completed=steps_completed, metrics=avg_metrics, batch_metrics=batch_metrics
+            )
+
+        self.context.reset_reducers()
+        self.training_metrics = []
+
+    def report_validation_metrics(self, steps_completed, metrics):
+        if self.is_chief:
+            # Check if op complete
+            searcher_metric = self._validate_searcher_metric(metrics)
+            if self.searcher_op and self._steps_remaining() < 1:
+                self.searcher_op.report_completed(searcher_metric)
+
+            self.core_context.train.report_validation_metrics(steps_completed, metrics)
+
+            if self.state.last_ckpt != self.steps_completed:
+                if self.ckpt_policy == "all" or (
+                    self.ckpt_policy == "best"
+                    and self.is_best_validation(now=searcher_metric,
+                                                before=self.core_context.train.get_experiment_best_validation())
+                ):
+                    self.checkpoint(already_exiting=False)
+
+    def is_best_validation(self, now: float, before: Optional[float]) -> bool:
+        if before is None:
+            return True
+        smaller_is_better = self.env.experiment_config["searcher"]["smaller_is_better"]
```

Comment:
This strikes me as something that is very experiment-config centric.  How will this look in the Trainer API for local training mode?

Reply:
We talked about this, and there really is no better way.

This would indicate that local training mode might want to provide a searcher_metric and smaller_is_better value though.